apis/python/src/tiledb/vector_search/ingestion.py

```diff
@@ -1637,6 +1693,69 @@ def create_ingestion_dag(
+                    # The number of data points to sample from each batch.
+                    samples_per_batch = 0 if num_batches == 0 else math.floor(training_sample_size / num_batches)
+                    # Calculate the remaining data points that need to be sampled (i.e. b/c the numbers were not perfectly divisible).
+                    remaining_samples = training_sample_size - (num_batches * samples_per_batch)
+                    # Calculate how many of the remaining data points to sample from each batch.
+                    extra_samples_per_batch = remaining_samples if num_batches == 0 else math.ceil(remaining_samples / num_batches)
+                    # Calculate at which index to stop sampling an extra_samples_per_batch from.
+                    idx_to_stop_adding_extra_samples = 0 if extra_samples_per_batch == 0 else remaining_samples / extra_samples_per_batch
+
+                    idx = 0
+                    num_sampled = 0
+                    for i in range(0, in_size, input_vectors_batch_size):
+                        # What vectors to read from the source_uri.
+                        start = i
+                        end = i + input_vectors_batch_size
+                        if end > size:
+                            end = size
+
+                        # How many of those vectors to sample in this batch.
+                        num_for_batch = samples_per_batch
+                        if idx < idx_to_stop_adding_extra_samples:
+                            num_for_batch += extra_samples_per_batch
+
+                        submit(
+                            random_sample_from_input_vectors,
+                            source_uri=source_uri,
+                            source_type=source_type,
+                            vector_type=vector_type,
+                            dimensions=dimensions,
+                            vector_start_pos=start,
+                            vector_end_pos=end,
+                            random_sample_size=num_for_batch,
+                            output_source_uri=training_source_uri,
+                            output_start_pos=num_sampled,
+                            output_end_pos=num_sampled + num_for_batch,
+                            config=config,
+                            verbose=verbose,
+                            name="read-random-sample-" + str(idx),
+                            resources={"cpu": str(threads), "memory": "12Gi"},
+                            image_name=DEFAULT_IMG_NAME,
+                        )
+                        num_sampled += num_for_batch
+                        idx += 1
+                    if num_sampled != training_sample_size:
+                        raise ValueError(f"The random sampling ran into an issue: num_sampled ({num_sampled}) != training_sample_size ({training_sample_size})")
+
                 if training_sample_size <= CENTRALISED_KMEANS_MAX_SAMPLE_SIZE:
```

Comment:
> I would suggest to also test this out with a larger input dataset in cloud to make sure that the distributed execution is working as expected.

Are you talking about a larger dataset so we hit the `else` condition here?
```
                if training_sample_size <= CENTRALISED_KMEANS_MAX_SAMPLE_SIZE:
                    centroids_node = submit(
                        centralised_kmeans,
                        index_group_uri=index_group_uri,
                        source_uri=source_uri,
                        source_type=source_type,
                        vector_type=vector_type,
                        partitions=partitions,
                        dimensions=dimensions,
                        training_sample_size=training_sample_size,
                        training_source_uri=training_source_uri,
                        training_source_type=training_source_type,
                        config=config,
                        verbose=verbose,
                        trace_id=trace_id,
                        use_sklearn=use_sklearn,
                        name="kmeans",
                        resources={"cpu": "8", "memory": "32Gi"},
                        image_name=DEFAULT_IMG_NAME,
                    )
                else:
                    internal_centroids_node = submit(
                        init_centroids,
                        source_uri=source_uri,
                        source_type=source_type,
                        vector_type=vector_type,
                        partitions=partitions,
                        dimensions=dimensions,
                        config=config,
                        verbose=verbose,
                        trace_id=trace_id,
                        name="init-centroids",
                        resources={"cpu": "1", "memory": "1Gi"},
                        image_name=DEFAULT_IMG_NAME,
                    )
```
If so, that code path currently is broken (the `NameError: name 'kmeans_predict' is not defined` error). So my proposal would be to just move forward with it working for `centralised_kmeans` and come back to test once that path is working. Or is there something else you're saying here that I'm missing?

Reply:
Thanks, sounds good. I've added a new cloud unit test and done some testing. I first tested `sift_10k`:
```
source_uri = "tiledb://TileDB-Inc/sift_10k"
training_sample_size = 1183
index = vs.ingest(
    index_type="IVF_FLAT",
    index_uri=index_uri,
    source_uri=source_uri,
    config=tiledb.cloud.Config().dict(),
    training_sampling_policy=vs.ingestion.TrainingSamplingPolicy.RANDOM,
    training_sample_size=training_sample_size,
    mode=Mode.BATCH,
)
```
Without `mode=Mode.BATCH` this passes. With `BATCH`, it works for the random sampling portion, though fails in `centralised_kmeans`:
```
test/test_cloud.py queries (100, 128)
[ingestion@random_sample_from_input_vectors] vector_start_pos: 0 vector_end_pos: 10000 batch: 20000000 random_sample_size: 1183 output_start_pos: 0
  sampled_vectors (1183, 128) float32 [  4.   4.   2.  71.  57.   0.   1.  21.   9.   0.   0.   7.  17.  19.
  43. 102.   0.   0.   0.   7.  42.  33.  23.  12.   4.   0.   0.   2.
   3.   0.   3.  13.  10.   4.  22.  61. 125.  41.  18.  10.   9.   3.
  23.  86.  32.   5.  15.  23.  15.   1.  10.  21.  10.   6.  12.  34.
  28.   3.  16.  18.   1.   0.   1.  17.  77.  19.  44.  68.  27.  19.
  35.  79.  18.   8.  29. 125.  72.   8.  21.  71.  53.  72.  40.  77.
  24.   0.   1.  25. 105.  28.  73.  89.   0.   0.   0.  17.  98.  24.
  59.  86.   3.   0.   0.  50.  73.  11.  34. 125.  53.   8.  11.  60.
  57. 125.  90.  38.   4.   0.   2.  12.  35.  41.  84. 125.   7.   0.
   0.   5.]
...
E           ImportError: cannot import name 'kmeans_fit' from 'tiledb.vector_search.module' (/opt/conda/lib/python3.9/site-packages/tiledb/vector_search/module.py)
 ```
 with `E           ImportError: cannot import name 'kmeans_fit' from 'tiledb.vector_search.module' (/opt/conda/lib/python3.9/site-packages/tiledb/vector_search/module.py)`

<img width="2032" alt="Screenshot 2024-01-09 at 1 41 32 PM" src="https://github.com/TileDB-Inc/TileDB-Vector-Search/assets/1396242/9fafb88e-4ad5-4a68-976a-ea2e9c7a4121">

I also tried with `ann_sift1b_raw_vectors_col_major` and it works up to `centralised_kmeans`. Here is the code:
```
        source_uri = "tiledb://TileDB-Inc/ann_sift1b_raw_vectors_col_major"
        max_sampling_tasks = 10
        training_sample_size = 1000000
        index = vs.ingest(
            index_type="IVF_FLAT",
            index_uri=index_uri,
            source_uri=source_uri,
            config=tiledb.cloud.Config().dict(),
            training_sampling_policy=vs.ingestion.TrainingSamplingPolicy.RANDOM,
            training_sample_size=training_sample_size,
            max_sampling_tasks=max_sampling_tasks,
            mode=Mode.BATCH,
            verbose=True,
        )
```
And here is the task graph and result: https://cloud.tiledb.com/activity/taskgraphs/TileDB-Inc/c25676e7-fcea-4903-9166-18a013e70da0

<img width="2032" alt="Screenshot 2024-01-10 at 12 52 18 PM" src="https://github.com/TileDB-Inc/TileDB-Vector-Search/assets/1396242/af9a87c3-7fc3-4820-9dd9-d5e639f5a0a7">

I can also see that within a Task (i.e. https://cloud.tiledb.com/activity/tasks/b7bafbb1-14bf-46d0-b3f3-251ca3275011) we are correctly looping over `batch`'s of data:
```
  vector_start_pos: 990000000 vector_end_pos: 1000000000 batch: 1000000 random_sample_size: 10000 output_start_pos: 990000
    start: 990000000, end: 991000000, percent_of_data_to_read: 0.1, num_to_sample: 1000
      [2024-01-10 10:36:22,616] [ingestion] [read_input_vectors] [DEBUG] Reading input vectors start_pos: 990000000, end_pos: 991000000
      vectors (1000000, 128) uint8
      sampled_vectors (1000, 128) uint8
    start: 991000000, end: 992000000, percent_of_data_to_read: 0.1, num_to_sample: 1000
      [2024-01-10 10:36:33,821] [ingestion] [read_input_vectors] [DEBUG] Reading input vectors start_pos: 991000000, end_pos: 992000000
      vectors (1000000, 128) uint8
      sampled_vectors (1000, 128) uint8
    start: 992000000, end: 993000000, percent_of_data_to_read: 0.1, num_to_sample: 1000
      [2024-01-10 10:36:49,034] [ingestion] [read_input_vectors] [DEBUG] Reading input vectors start_pos: 992000000, end_pos: 993000000
      vectors (1000000, 128) uint8
      sampled_vectors (1000, 128) uint8
    start: 993000000, end: 994000000, percent_of_data_to_read: 0.1, num_to_sample: 1000
      [2024-01-10 10:37:04,125] [ingestion] [read_input_vectors] [DEBUG] Reading input vectors start_pos: 993000000, end_pos: 994000000
      vectors (1000000, 128) uint8
      sampled_vectors (1000, 128) uint8
    start: 994000000, end: 995000000, percent_of_data_to_read: 0.1, num_to_sample: 1000
      [2024-01-10 10:37:14,100] [ingestion] [read_input_vectors] [DEBUG] Reading input vectors start_pos: 994000000, end_pos: 995000000
      vectors (1000000, 128) uint8
      sampled_vectors (1000, 128) uint8
    start: 995000000, end: 996000000, percent_of_data_to_read: 0.1, num_to_sample: 1000
      [2024-01-10 10:37:30,378] [ingestion] [read_input_vectors] [DEBUG] Reading input vectors start_pos: 995000000, end_pos: 996000000
      vectors (1000000, 128) uint8
      sampled_vectors (1000, 128) uint8
    start: 996000000, end: 997000000, percent_of_data_to_read: 0.1, num_to_sample: 1000
      [2024-01-10 10:37:40,283] [ingestion] [read_input_vectors] [DEBUG] Reading input vectors start_pos: 996000000, end_pos: 997000000
      vectors (1000000, 128) uint8
      sampled_vectors (1000, 128) uint8
    start: 997000000, end: 998000000, percent_of_data_to_read: 0.1, num_to_sample: 1000
      [2024-01-10 10:37:52,707] [ingestion] [read_input_vectors] [DEBUG] Reading input vectors start_pos: 997000000, end_pos: 998000000
      vectors (1000000, 128) uint8
      sampled_vectors (1000, 128) uint8
    start: 998000000, end: 999000000, percent_of_data_to_read: 0.1, num_to_sample: 1000
      [2024-01-10 10:38:03,371] [ingestion] [read_input_vectors] [DEBUG] Reading input vectors start_pos: 998000000, end_pos: 999000000
      vectors (1000000, 128) uint8
      sampled_vectors (1000, 128) uint8
    start: 999000000, end: 1000000000, percent_of_data_to_read: 0.1, num_to_sample: 1000
      [2024-01-10 10:38:14,797] [ingestion] [read_input_vectors] [DEBUG] Reading input vectors start_pos: 999000000, end_pos: 1000000000
      vectors (1000000, 128) uint8
      sampled_vectors (1000, 128) uint8
```

Reply:
Thanks for testing! 

For the kmeans error, you could try to use sklearn kmeans by setting `use_sklearn`.